src/gui/updater/appimageupdater.cpp

```diff
@@ -0,0 +1,206 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include <QMessageBox>
+#include <QSettings>
+#include <QTimer>
+#include <appimage/update.h>
+
+#include "appimageupdater.h"
+#include "common/version.h"
+#include "theme.h"
+#include "updater_private.h"
+// SettingsDialog is apparently forward-declared somewhere, need to include the header to make the compilation work
+#include "libsync/configfile.h"
+#include "settingsdialog.h"
+
+#include "appimageupdateavailabledialog.h"
+#include "application.h"
+
+using namespace OCC;
+
+namespace {
+
+/**
+ * libappimageupdate uses exceptions, but the client does not
+ * This little shim adapts the interface to one usable within this project
+ */
+class AppImageUpdaterShim : public QObject
+{
+    Q_OBJECT
+
+private:
+    explicit AppImageUpdaterShim(const QString &zsyncFileUrl, QObject *parent = nullptr)
+        : QObject(parent)
+        , _updater(Utility::appImageLocation().toStdString(), true)
+    {
+        QString updateInformation(QStringLiteral("zsync|") + zsyncFileUrl);
+        _updater.setUpdateInformation(updateInformation.toStdString());
+    }
+
+    void _logStatusMessages()
+    {
+        std::string currentStatusMessage;
+
+        while (_updater.nextStatusMessage(currentStatusMessage)) {
+            qCInfo(lcUpdater) << "AppImageUpdate:" << QString::fromStdString(currentStatusMessage);
+        }
+    }
+
+public:
+    static AppImageUpdaterShim *makeInstance(const QString &updateInformation, QObject *parent)
+    {
+        try {
+            return new AppImageUpdaterShim(updateInformation, parent);
+        } catch (const std::exception &e) {
+            qCCritical(lcUpdater) << "Failed to create updater shim:" << e.what();
+            return nullptr;
+        }
+    }
+
+    bool isUpdateAvailable() noexcept
+    {
+        try {
+            bool updateAvailable;
+
+            if (!_updater.checkForChanges(updateAvailable)) {
+                _logStatusMessages();
+                return false;
+            }
+
+            _logStatusMessages();
+            return updateAvailable;
+        } catch (const std::exception &e) {
+            _logStatusMessages();
+            qCCritical(lcUpdater) << "Checking for update failed:" << e.what();
+            return false;
+        }
+    }
+
+    void startUpdateInBackground() noexcept
+    {
+        // monitor progress and log status messages
+        auto *timer = new QTimer(this);
+
+        timer->setInterval(100);
+
+        connect(timer, &QTimer::timeout, this, [=]() {
+            _logStatusMessages();
+
+            if (!_updater.isDone()) {
+                emit finished(!_updater.hasError());
+            }
+
+            timer->stop();
```

Comment:
We always stop after 100ms?
I was expecting some polling here?

Reply:
Indeed, this makes no sense. The `if` above is broken, too. I'll fix it.